crates/accelerate/src/synthesis/cnot_synthesis.rs

```diff
@@ -0,0 +1,163 @@
+    // let mut mat_t = mat.t().to_owned();
+    let upper_cnots = lower_cnot_synth(&mut mat, &(section_size as usize), &true);
+
+    // iterator over the gates
+    let instructions = upper_cnots
+        .iter()
+        .map(|(i, j)| (*j, *i))
+        .chain(lower_cnots.into_iter().rev())
+        .map(|(ctrl, target)| {
+            (
+                StandardGate::CXGate,
+                smallvec![],
+                smallvec![Qubit(ctrl as u32), Qubit(target as u32)],
+            )
+        });
+
+    Ok(
+        CircuitData::from_standard_gates(py, num_qubits as u32, instructions, Param::Float(0.0))
+            .expect("Something went sideways in Qiskit's Python realm!"),
+    )
+}
+
+/// instead of transposing the matrix, allow setting the axis in this function
+fn lower_cnot_synth(
+    matrix: &mut Array2<bool>,
+    section_size: &usize,
+    transpose: &bool,
+) -> Vec<(usize, usize)> {
+    // The vector of CNOTs to be applied. Called ``circuit`` here for consistency with the paper.
+    let mut circuit: Vec<(usize, usize)> = Vec::new();
+    let cutoff = 1;
+
+    // to apply to the transposed matrix, we can just set axis = 1
+    let row_axis = if *transpose { Axis(1) } else { Axis(0) };
+
+    // get number of columns (same as rows) and the number of sections
+    let n = matrix.raw_dim()[0];
+    let num_sections = n / section_size + 1;
+
+    // iterate over the columns
+    for section in 1..num_sections {
+        // store sub section row patterns here, which we saw already
+        let mut patterns: HashMap<Array1<bool>, usize> = HashMap::new();
+        let section_slice = s![(section - 1) * section_size..cmp::min(section * section_size, n)];
+
+        // iterate over the rows (note we only iterate from the diagonal downwards)
+        for row_idx in (section - 1) * section_size..n {
+            // we need to keep track of the rows we saw already, called ``pattern`` here
+            let pattern: Array1<bool> = matrix
+                .index_axis(row_axis, row_idx)
+                .slice(section_slice)
+                .to_owned();
+
+            // skip if the row is empty (i.e. all elements are false)
+            if pattern.iter().any(|&el| el) {
+                if patterns.contains_key(&pattern) {
+                    // store CX location
+                    circuit.push((patterns[&pattern], row_idx));
+                    // remove the row
+                    _add(matrix, transpose, patterns[&pattern], row_idx);
+                } else {
+                    // if we have not seen this pattern yet, keep track of it
+                    patterns.insert(pattern, row_idx);
+                }
+            }
+        }
+
+        // gaussian eliminate the remainder of the section
+        for col_idx in (section - 1) * section_size..section * section_size {
+            let mut diag_el = matrix[[col_idx, col_idx]];
+
+            for r in col_idx + 1..n {
```

Comment:
I think this can be done using iterators like in the usual gaussian elimination in #12456

Reply:
I could get it more or less to work without the "backreduce" in the end of the iteration (L147 and following). But the backreduce accesses the pivot row repeatedly and Rust is complaining about multiple mutable borrows in the iterator... If you have any more ideas I'm happy to try more 🙂